Add multiplot support and refactor plot.hpp

HorstBaerbel opened the pull request:
This, is rather big, but please take your time to review this.
* Add support for GNUplot [multiplots](http://gnuplot.sourceforge.net/docs_4.2/node203.html). This adds the class multiplot that lets you add() plots to it and then show() or save() as with plot. See the example-multiplot.cpp.
* Remove all std::ofstream and FILE* members from plot / multiplot. It is not a good idea to keep those around longer than you need them. It also made the plot class non-copyable (copying an allocated pointer / file pointer is usually no-no). The pipe is closed directly after popen() using a pclose(), which waits for the termination of the sub-process.
* Files are closed when they have been written to. The data file is reopened for every plot::draw() call and closed after writing data, which eliminates the need to flush() it. The script file is closed too after writing.
* The plot index counter has been reworked. There's now an id() function that returns the plot "number" the plot got on creation.
* GNUplot specific outputs have been extracted as functions and moved to util.hpp.
* I removed all line breaks inside functions. They make code hard to read imo and there's lots of comments for "visual line-breaks" too.
* Update README a bit.

allanleal commented on sciplot/multiplot.hpp:
```diff
@@ -0,0 +1,227 @@
+    using namespace internal;
+
+    /// The class used to define multipleplots in one canvas, show them in a pop-up window or save them to a file.
+    class multiplot : public plot
```
Is inheritance here really needed?

allanleal commented on sciplot/multiplot.hpp:
```diff
@@ -0,0 +1,227 @@
+        auto id() const -> std::size_t;
+
+        /// Counter of how many plot / singleplot objects have been instanciated in the application
+        static std::size_t m_counter;
```
How do we use this `static` counter? Is it possible not to use `static`?

HorstBaerbel replied:
Well, some way the plot / multiplot instances have to be counted to not overwrite files and to connect script to data files. Callers don't care about that detail, so this way it is hidden from them. We could remove the id() method and just use m_id (yagni).

HorstBaerbel replied:
I removed the id() function now and [made the counter private again](https://github.com/sciplot/sciplot/pull/10/commits/4d0aaf53100658a5c9649dfcb8f56a0350f14883). If we use [random temporary files](https://en.cppreference.com/w/cpp/io/c/tmpnam), we probably could get rid of it.

allanleal replied:
OK, now I remember the use of this static counter. The disadvantage of using random temp files is that it is not clear which generated files correspond to the sequence of plot calls. I mean, if two plots are generated, their order indication in the file name may be useful to some users, for further inspecting the corresponding generated gnuplot scripts (check my comment here in the other issue you created #12 ). In the future, we could have different options on how files are generated, name schemes, and cleaning procedure for the generated gnuplot scripts.

HorstBaerbel replied:
Let's take the discussion about this to #12

allanleal commented on sciplot/plot.hpp:
```diff
@@ -42,334 +42,295 @@
-
-/// The class used to define the plots, show in a pop-up window and save to a file.
-class plot
+namespace sciplot
```
What about no indentation inside `namespace`:

~~~cpp
namespace sciplot {
...
} // namespace sciplot

HorstBaerbel replied:
Ok with me. I rely on auto-formatting, so I could check if I can prepare a .clang-format file that does that. Should I commit that so we can use it in the CI too?

HorstBaerbel replied:
See 23664c8 to check if this is ok with you.

allanleal replied:
This is all fine as long as we can slightly adapt it in the future if needed! ;)

HorstBaerbel commented:
About the multiplit->plot inheritance: Good catch! Definitely not needed! Wtf did it to there?!

HorstBaerbel commented on sciplot/multiplot.hpp:
```diff
@@ -0,0 +1,224 @@
+namespace sciplot
+{
+
+    using namespace internal;
```
I'm not sure I understand your comment. I haven't given this much thought. I basically stole from plot.hpp.
Imo it is bad practice to using namespace in a header file (if find it ok in source files), so maybe we should clean this up...

HorstBaerbel replied:
Why are classes like axislabelspecs sciplot::**internal** btw when the user uses them?!

HorstBaerbel replied:
This is fixed now in 9e58f1d

HorstBaerbel commented:
I've removed the "using namespace internal" from plot.hpp and multiplot.hpp, so users shouldn't come into contact with sciplot::internal anymore. As a library developer you have to write more, but imo it's better that the lib looks good on the user side and hides it internals.  
I've moved all the functions formatting or writing gnuplot data or running the scripts to a "gnuplot" namespace. They were prepended with "gnuplot" anyway...

HorstBaerbel commented:
I know I can merge now myself (thanks for the trust btw), but I'd definitely like your opinion on changes...

HorstBaerbel commented:
Added a .clang-format file that can be used with the clang-format executable, but also with the Clang-format extension (xaver.clang-format) in VS Code. Turn on "Editor: Format on save" in the VS Code settings and everything is always properly formatted. If you don't like something, change the .clang-format file.
Changes (imo pretty minor):
* namespace definitions have a line break before the brace
* visibility statements (e.g. private:) are indented 2 spaces
* constructor initializer lists are indented 4 spaces
* functions longer than one line (or 80 characters?) have line breaks added

allanleal commented on examples/example-multiplot.cpp:
```diff
@@ -0,0 +1,57 @@
+{
+    // Create 4 different plots
+    const vec x = linspace(0.0, 5.0, 200);
+    plot p0;
```
I think this nice example would be better presented if each block of plot commands are separated by empty lines.

~~~cpp
plot p0;
p0.draw(x, std::sin(x)).title("sin(x)");
p0.draw(x, std::cos(x)).title("cos(x)");

plot p1;
p1.draw(x, std::cos(x)).title("cos(x)");

plot p2;
p2.draw(x, std::tan(x)).title("tan(x)");

plot p3;
p3.draw(x, std::sqrt(x)).title("sqrt(x)");
~~~

allanleal commented on sciplot/specs/gridspecs.hpp:
```diff
@@ -30,68 +30,77 @@
 
     /// Return a grid specification object for configuring grid lines along minor ytics.
-    auto mytics() -> gridspecsbase& { return _gridminor("mytics"); }
+    auto mytics() -> gridspecsbase & { return _gridminor("mytics"); }
```
I think it is more common to have `&` attached to the type (i.e. `gridspecsbase&`).

HorstBaerbel replied:
This is somewhat of a religious thing. [Googles convention](https://google.github.io/styleguide/cppguide.html#Pointer_and_Reference_Expressions) is "use one convention per file and stick with it". Stackoverflow answers very. I don't care much, so I just set "PointerAlignment: Left". Will refomat the files.

HorstBaerbel commented on README.md:
````diff
@@ -5,49 +5,50 @@
     plt.save("figure.pdf");
 }
-~~~
+```
````
[Googles convention ](https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace)basically says to add vertical space where it makes sense. I'm all for dense code, so as much code as possible (without being illegible) fits on one page (horizontally and vertically).
I'd suggest vertical space where there's a logical break in code flow, e.g. where you'd make a pargraph in text, but not after every line. Like:
```c++
...
// Do complicated thing A
statement();
statement();

// Now the result from A is here, now do B
statement();
statement();
...
```
Would that be ok with you? I can format plot.hpp and multiplot.hpp in that way.

HorstBaerbel replied:
Fixed in d6b15f3

allanleal commented:
OK, not a consistent convention then! :)

On Mon, Jun 29, 2020 at 4:02 PM Bim <notifications@github.com> wrote:

> *@HorstBaerbel* commented on this pull request.
> ------------------------------
>
> In sciplot/specs/gridspecs.hpp
> <https://github.com/sciplot/sciplot/pull/10#discussion_r446994990>:
>
> >
>      /// Return a grid specification object for configuring grid lines along minor ytics.
> -    auto mytics() -> gridspecsbase& { return _gridminor("mytics"); }
> +    auto mytics() -> gridspecsbase & { return _gridminor("mytics"); }
>
> This is somewhat of a religious thing. Googles convention
> <https://google.github.io/styleguide/cppguide.html#Pointer_and_Reference_Expressions>
> is "use one convention per file and stick with it". Stackoverflow answers
> very. I don't care much, so I just set "PointerAlignment: Left". Will
> refomat the files.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sciplot/sciplot/pull/10#discussion_r446994990>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMOINBLUXRIC34XUOQXKXLRZCNFZANCNFSM4OJFFS4A>
> .
>

allanleal commented:
This is fine - vertical space when we start a new block of code as in the
example above.

On Mon, Jun 29, 2020 at 3:47 PM Bim <notifications@github.com> wrote:

> *@HorstBaerbel* commented on this pull request.
> ------------------------------
>
> In README.md
> <https://github.com/sciplot/sciplot/pull/10#discussion_r446984515>:
>
> >      plt.save("figure.pdf");
>  }
> -~~~
> +```
>
> Googles convention
> <https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace>basically
> says to add vertical space where it makes sense. I'm all for dense code, so
> as much code as possible (without being illegible) fits on one page
> (horizontally and vertically).
> I'd suggest vertical space where there's a logical break in code flow,
> e.g. where you'd make a pargraph in text, but not after every line. Like:
>
> ...// Do complicated thing Astatement();statement();
> // Now the result from A is here, now do Bstatement();statement();
> ...
>
> Would that be ok with you? I can format plot.hpp and multiplot.hpp in that
> way.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sciplot/sciplot/pull/10#discussion_r446984515>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMOINBWU3LCFJS6PJZO7JTRZCLP3ANCNFSM4OJFFS4A>
> .
>

allanleal commented:
Sounds good!

On Mon, Jun 29, 2020 at 2:55 PM Bim <notifications@github.com> wrote:

> *@HorstBaerbel* commented on this pull request.
> ------------------------------
>
> In sciplot/multiplot.hpp
> <https://github.com/sciplot/sciplot/pull/10#discussion_r446948476>:
>
> > +        /// Show the plot in a pop-up window.
> +        auto show() -> void;
> +
> +        /// Save the plot in a file, with its extension defining the figure format.
> +        /// The extension of the file name determines the format of the figure.
> +        /// The supported figure formats are: `pdf`, `eps`, `svg`, `png`, and `jpeg`.
> +        /// Thus, to save a plot in `png` format, choose a file name with a `.png`
> +        /// file extension as in `fig.png`.
> +        auto save(const std::string &filename) -> void;
> +
> +    protected:
> +        /// Plot id derived from m_counter upon construction
> +        auto id() const -> std::size_t;
> +
> +        /// Counter of how many plot / singleplot objects have been instanciated in the application
> +        static std::size_t m_counter;
>
> Let's take the discussion about this to #12
> <https://github.com/sciplot/sciplot/issues/12>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sciplot/sciplot/pull/10#discussion_r446948476>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMOINGJFVBHDIVRRBOHJY3RZCFLNANCNFSM4OJFFS4A>
> .
>

allanleal commented:
Please request my review when you are done and I'll just have a quick look
at this and we merge this PR.

On Mon, Jun 29, 2020 at 5:19 PM Bim <notifications@github.com> wrote:

> *@HorstBaerbel* commented on this pull request.
> ------------------------------
>
> In README.md
> <https://github.com/sciplot/sciplot/pull/10#discussion_r447052207>:
>
> >      plt.save("figure.pdf");
>  }
> -~~~
> +```
>
> Fixed in d6b15f3
> <https://github.com/sciplot/sciplot/commit/d6b15f35a31a9434228deb987ff74cc644178ef1>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sciplot/sciplot/pull/10#discussion_r447052207>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMOINAAYRR7CPWENEZLSZTRZCWJXANCNFSM4OJFFS4A>
> .
>

allanleal commented:
Great. Thank you.


On Mon, Jun 29, 2020 at 5:47 PM Bim <notifications@github.com> wrote:

> Merged #10 <https://github.com/sciplot/sciplot/pull/10> into master.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sciplot/sciplot/pull/10#event-3493383607>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMOINFALXKDHUWC2LJ76RDRZCZQJANCNFSM4OJFFS4A>
> .
>

HorstBaerbel commented:
Ah, sorry. I though I'd merge as I addressed all the thing you pointed out to. If there's stuff left over, we can always do this in the next PR...

allanleal commented:
No worries!

On Mon, Jun 29, 2020 at 5:50 PM Bim <notifications@github.com> wrote:

> Ah, sorry. I thought I'd merge as I addressed all the things you pointed
> out to. If there's stuff left over, we can always do this in the next PR...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sciplot/sciplot/pull/10#issuecomment-651206419>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMOINE3DZHIBGZ3K5LFXEDRZCZ43ANCNFSM4OJFFS4A>
> .
>
